thinc/backends/ops.py

```diff
@@ -227,11 +259,11 @@ def affine(self, X: Floats2d, W: Floats2d, b: Floats1d) -> Floats2d:
 
     def flatten(
         self,
-        X: Sequence[ArrayT],
+        X: List[ArrayTXd_co],
```

Comment:
So there are two ways of achieving a covariant collection: by using covariant types or by using read-only collections like `Sequence`. We have to use the covariant type here because it's also the return type. So `Sequence` is no longer necessary for covariance, although there's no reason we couldn't continue using it if we wanted to — I guess `List` just seemed more consistent with the rest of the module/library.

There's no problem reassigning X to a new `Sequence`, you just can't change the contents of an existing `Sequence`.

Do you think I should change the code here? I was generally trying to avoid doing so and just stick to changing the types because of the late stage in the development cycle for the release.

Reply:
I agree it's probably best to not touch the code too much and keep this PR focused on the types. The reason I brought it up because I think we should use `Sequence` whenever is applicable - i.e. when the variable isn't changed by the function. That should be the case here, so I would switch it back. There's actually similar places in `loss.py` where we should be switching to `Sequence` as well, cf https://github.com/explosion/spaCy/pull/10231#discussion_r809022086

Reply:
Sorry, I'm still hung up on this. What is the use-case we're supporting to use the covariant typing here?
I'm trying to wrap my head around this. Can you give an example of a snippet of code that wasn't working with the original type annotations?

Reply:
(Btw I've changed it back to `Sequence` here and also changed `List` to `Sequence` throughout `loss.py`.)

The original definition was 

```
def flatten(
        self,
        X: Sequence[ArrayT],
        dtype: Optional[DTypes] = None,
        pad: int = 0,
        ndim_if_empty: int = 2,
    ) -> ArrayT:
```

`ArrayT` had the same definition that `ArrayTXd` now does:

`ArrayTXd = TypeVar("ArrayTXd", bound=ArrayXd)
`
meaning that if you called the method

`flattened = ops.flatten(...
`
and wanted to give `flattened` a type, you would have had to assign `Union[Ints1d, Ints2d, Ints3d, Ints4d, Floats1d, Floats2d, Floats3d, Floats4d]` regardless of what you passed in.

Using the covariant type here, on the other hand, binds the return type to the type within the list that is passed in.

Reply:
I can't seem to reproduce that though - which is where I remain confused.

Let's say we'd have the definition without covariance:
```
    def flatten(
        self,
        X: Sequence[ArrayTXd],
        ....
    ) -> ArrayTXd:
```

Now I'm trying with this code:
```
def call_flatten(F_3: Floats3d):
    ops = get_current_ops()
    my_list = [F_3, F_3]
    flattened: Floats3d = ops.flatten(my_list)
    print(flattened)


if __name__ == "__main__":
    F_3: Floats3d = cast(Floats3d, numpy.zeros((2, 2, 2), dtype="f"))
    call_flatten(F_3)
```
That seems to run and type check just fine? i.e. the return type is the same as the input type, and can be typed as such explicitely?

Reply:
I guess the fact that a `Sequence` was in use at the start had somehow tricked me into thinking this must be somewhere where covariance was necessary.